Add WIP syncing functionality

broodroosterdev opened the pull request.

broodroosterdev committed:
notes: Implement get_it dependency injection

broodroosterdev committed:
notes: Add NoteController for communication with the sync api

broodroosterdev committed:
notes: Add NoteController for communication with the sync api

broodroosterdev committed:
notes: Match functions of note_controller with its interface

The note_interface didnt really match what the API would ask for so the function signatures needed to be changed.

broodroosterdev committed:
notes: Use uuid's as note id

broodroosterdev committed:
notes: Add SyncNote class in preparation of syncing

broodroosterdev committed:
notes: Add SyncNote class in preparation of syncing

broodroosterdev committed:
notes: Add SyncRoutine class for syncing notes

broodroosterdev committed:
notes: Sync when user pulls down on main page

broodroosterdev committed:
notes: Login at startup to retrieve JWT tokens

dvd-bnc commented on lib/data/dao/note_helper.dart:
```diff
@@ -13,15 +13,23 @@ class NoteHelper extends DatabaseAccessor<AppDatabase> with _$NoteHelperMixin {
       case ReturnMode.NORMAL:
         return (select(notes)
-              ..where((table) => table.archived.not() & table.deleted.not()))
+              ..where((table) => table.archived.not() & table.deleted.not() & table.id.contains("-synced").not()))
```
you may prefer to use endsWith instead of contains

broodroosterdev replied:
You are totally right. This is really useful feedback, thanks!

broodroosterdev replied:
Apparently the type of id is GeneratedTextColumn, which doesnt have all the methods a string has. Looks like it will have to keep using contains, even though endsWith would be better.

dvd-bnc replied:
i can look into it dw

dvd-bnc commented on lib/data/model/sync_note.dart:
```diff
@@ -0,0 +1,66 @@
+    bool deleted = note.deleted;
+    bool archived = note.archived;
+    bool synced = note.synced;
+    return SyncNote(id, title, content, styleJson, starred, creationDate, lastModifyDate, color, images, list, listContent, reminders, hideContent, lockNote, usesBiometrics, deleted, archived, synced);
```
oof run flutter format on this file, that line is too long for sure

broodroosterdev replied:
Just did and yeah, way too long.

dvd-bnc commented on lib/routes/main_page.dart:
```diff
@@ -314,4 +318,7 @@ class _MainPageState extends State<MainPage> with TickerProviderStateMixin {
     }
   }
+  Future<void> sync() async {
+    await SyncRoutine().syncNotes();
```
may be better to add the SyncRoutine instance to the getit locator as a singleton so you dont have to always reinstance it

broodroosterdev replied:
I agree, i will add it to the locator.

dvd-bnc commented on lib/internal/sync/sync_routine.dart:
```diff
@@ -0,0 +1,112 @@
+      return;
+    }
+    await updateLists();
+    NoteController noteController = NoteController();
```
also this here, might want to put it in the locator

broodroosterdev replied:
Actually, now that i think about it, the notecontroller should just use static functions since it doesnt store state or anything

broodroosterdev replied:
For some reason i cant add static methods to an interface so i just added it to the locator

dvd-bnc replied:
can look into it, there might be another way to do so (or no way at all)

dvd-bnc replied:
k so you usually cant, but there really isnt a reason for keeping the abstract class, makes no sense cuz we dont have different implementations for any controller, imo you could just delete the *Interface classes and keep the controllers only with the static methods

dvd-bnc commented on lib/main.dart:
```diff
@@ -16,6 +17,8 @@ main() async {
   db = AppDatabase(constructDb());
   setupLocator();
   await locator<Preferences>().loadData();
+  final account = AccountController();
```
now i know this is only temporary, but still we should move every *Controller class to the locator for better access and for not having a shit ton of instances all around the app

broodroosterdev replied:
I'll add the AccountController to it as well

broodroosterdev committed:
notes: Add controllers and SyncRoutine to locator

This commit adds the following classes to the locator: NoteController, AccountController and SyncRoutine.

broodroosterdev committed:
notes: Add controllers and SyncRoutine to locator

This commit adds the following classes to the locator: NoteController, AccountController and SyncRoutine.

dvd-bnc commented on lib/data/model/sync_note.dart:
```diff
@@ -25,14 +25,27 @@ class SyncNote{
+      this.uses_biometrics,
+      this.deleted,
+      this.archived,
       this.synced);
```
add comma at the end of synced and reformat

broodroosterdev replied:
check latest commit

dvd-bnc commented on lib/locator.dart:
```diff
@@ -1,15 +1,19 @@
   locator.registerSingleton<Preferences>(Preferences());
   locator.registerLazySingleton<NoteHelper>(() => NoteHelper(db));
+  locator.registerLazySingleton<SyncRoutine>(() => SyncRoutine());
+  locator.registerLazySingleton(() => NoteController());
```
any reason why you didn't declare the types here for the controllers? (like registerLazySingleton<NoteController>(() ...

broodroosterdev replied:
Typo, my bad sorry

broodroosterdev replied:
Check latest commit

broodroosterdev committed:
notes: Fix locator bindings

broodroosterdev committed:
notes: Fix locator bindings

dvd-bnc commented on lib/internal/sync/controller/note_controller.dart:
```diff
@@ -12,44 +12,40 @@ import 'package:provider/provider.dart';
 import 'package:http/http.dart' as http;
```
why did this file get unformatted this commit

broodroosterdev replied:
Accidentally reformatted all files, reverting it probably changed it. I'll format it again and commit it. Honestly you should really squash all these commits into one. I am not proud of this mess of a tree.

broodroosterdev replied:
latest commit should fix it

dvd-bnc replied:
will squash no probs

broodroosterdev commented on lib/routes/note_page.dart:
```diff
@@ -98,15 +100,8 @@ class _NotePageState extends State<NotePage> {
-    }
-
-    if (note.id == null) note = note.copyWith(id: (lastNote?.id ?? 0) + 1);
+    var uuid = Uuid();
```
Might need to add this to locator

broodroosterdev replied:
Did in [newest commit](https://github.com/HrX03/PotatoNotes/pull/34/commits/c6a0c51e3f9efb3078be64c62a013a3719ec0a82)

broodroosterdev committed:
notes: Fix formatting of NoteController

broodroosterdev committed:
notes: Add Uuid to locator

broodroosterdev committed:
notes: Remove the SyncNote class

broodroosterdev committed:
notes: Remove the SyncNote class

dvd-bnc commented on lib/internal/sync/controller/note_controller.dart:
```diff
@@ -20,8 +19,7 @@ class NoteController implements NoteInterface {
   String toJson(Note note) {
-    SyncNote savingNote = SyncNote.fromNote(note);
-    return json.encode(savingNote);
+    return json.encode(Utils.toSyncMap(note));
```
why does this even exist at all

broodroosterdev replied:
I will remove it, it used to be more lines

dvd-bnc commented on lib/internal/utils.dart:
```diff
@@ -16,7 +22,44 @@ class Utils {
+      }
+      newMap.putIfAbsent(newKey, () => newValue);
+    });
+    return newMap;
```
if you always need the encoded json just put it on here, maybe with a bool param to omit encoding

broodroosterdev replied:
I'll prolly just make a small function for it like toSyncJson(Note note)

broodroosterdev committed:
notes: Remove the toJson function

broodroosterdev committed:
notes: Clear lists when finishing syncing. Fix notes not getting synced

Fixed bug which would add deleted notes to the app database. This fix will clear the added, updated and deleted list when finishing syncing.
This commit will also make sure that however you change the note (especially when not opening the note e.g. changing color, lock status etc.), it wont prevent the note getting synced.

broodroosterdev committed:
notes: Clear lists when finishing syncing. Fix notes not getting synced

Fixed bug which would add deleted notes to the app database. This fix will clear the added, updated and deleted list when finishing syncing.
This commit will also make sure that however you change the note (especially when not opening the note e.g. changing color, lock status etc.), it wont prevent the note getting synced.

dvd-bnc commented on lib/internal/sync/sync_routine.dart:
```diff
@@ -34,31 +31,46 @@ class SyncRoutine {
-      if (result == true) {
-        deleteSynced(note);
+  void syncNotes() {
+    checkOnlineStatus().then((status) {
```
I usually don't trust .then(), might be better to use async/await as it doesn't really change the flow in here and can be done easily

dvd-bnc commented on lib/routes/main_page.dart:
```diff
@@ -319,6 +319,8 @@ class _MainPageState extends State<MainPage> with TickerProviderStateMixin {
   Future<void> sync() async {
-    await locator<SyncRoutine>().syncNotes();
+    Future(() {
+      locator<SyncRoutine>().syncNotes();
```
What the hell is this, why did you do this, the one before was the same but much cleaner

dvd-bnc commented on lib/routes/main_page.dart:
```diff
@@ -319,6 +319,8 @@ class _MainPageState extends State<MainPage> with TickerProviderStateMixin {
     }
   }
   Future<void> sync() async {
```
Also you can remove Future<void> and use directly void

broodroosterdev replied:
Just tried removing Future but it gives an error since the onRefresh: property only accepts Future<void> functions

dvd-bnc replied:
Right, then leave as is

dvd-bnc commented on lib/widget/selection_bar.dart:
```diff
@@ -53,10 +53,10 @@ class SelectionBar extends StatelessWidget {
             for (int i = 0; i < selectionList.length; i++) {
               if (anyNotStarred)
-                await helper.saveNote(selectionList[i].copyWith(starred: true));
+                await helper.saveNote(selectionList[i].copyWith(starred: true, synced: false, lastModifyDate: DateTime.now()));
```
Might wanna format this file as the lines are too long

dvd-bnc commented on lib/routes/settings_page.dart:
```diff
@@ -89,7 +89,7 @@ class _SettingsPageState extends State<SettingsPage> {
                         for (int i = 0; i < notes.length; i++) {
                           await helper
-                              .saveNote(notes[i].copyWith(lockNote: false));
+                              .saveNote(notes[i].copyWith(lockNote: false, synced: false, lastModifyDate: DateTime.now()));
```
Requires format

broodroosterdev commented:
Yeah i agree, the whole Future(().. thing isn't necessary. I thought at first it might interfere when its done asynchronously but it wont.

broodroosterdev committed:
notes: Clean up code of syncNotes and fix formatting

broodroosterdev committed:
notes: Clean up code of syncNotes and fix formatting

broodroosterdev committed:
notes: Download notes when syncing

Syncing will now download notes from the api at the end of the sync procedure. It will use a last_updated timestamp to ensure only updated notes get sent back, thus saving on bandwidth.

broodroosterdev committed:
Merge upstream/rewrite into rewrite

broodroosterdev commented:
The build failing is a result of when i upped the sdk version to get App-auth. Since that is no longer needed i'll try to remove it and change it back to the original version

broodroosterdev committed:
notes: Change to certain gradle version

This will allow building with the default NDK version

broodroosterdev committed:
notes: Add proper error handling to account_controller

broodroosterdev committed:
notes: Change to a working gradle version

broodroosterdev committed:
notes: Update API routes and improve error handling

dvd-bnc commented:
This PR has become a big mess and was cleaned and adapted on sync branch, closing this
